pkg/prometheus/promcfg_test.go

```diff
@@ -6444,3 +6444,500 @@ func TestTracingConfig(t *testing.T) {
 		})
 	}
 }
+
+func TestServiceMonitorScrapeClass(t *testing.T) {
+	// test table for scrape class
+	testCases := []struct {
+		name        string
+		scrapeClass *monitoringv1.ScrapeClass
+		golden      string
+		expectedErr bool
+	}{
+		{
+			name:        "Service Monitor without Scrape Class",
+			golden:      "serviceMonitorWithoutScrapeClass.golden",
+			scrapeClass: nil,
+		},
+		{
+			name:   "Service Monitor Scrape Class With TLS Config",
+			golden: "serviceMonitorScrapeClassTlsConfig.golden",
+			scrapeClass: &monitoringv1.ScrapeClass{
+				Name: "test-tls-scrape-class",
+				TLSConfig: &monitoringv1.TLSConfig{
+					CAFile:   "/etc/prometheus/secrets/ca.crt",
+					CertFile: "/etc/prometheus/secrets/tls.crt",
+					KeyFile:  "/etc/prometheus/secrets/tls.key",
+				},
+			},
+		},
+		{
+			name:   "Service Monitor Scrape Class With Authorization Config",
+			golden: "serviceMonitorScrapeClassAuthorizationConfig.golden",
+			scrapeClass: &monitoringv1.ScrapeClass{
+				Name: "test-authorization-scrape-class",
+				Authorization: &monitoringv1.Authorization{
+					CredentialsFile: "/etc/prometheus/secrets/credentials",
+				},
+			},
+		},
+		{
+			name:   "Service Monitor Default Scrape Class",
+			golden: "serviceMonitorDefaultScrapeClass.golden",
+			scrapeClass: &monitoringv1.ScrapeClass{
+				Name:    "test-default-scrape-class",
+				Default: ptr.To(true),
+				Authorization: &monitoringv1.Authorization{
+					CredentialsFile: "/etc/prometheus/secrets/default",
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		p := defaultPrometheus()
+
+		monitor := monitoringv1.ServiceMonitor{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "testservicemonitor1",
+				Namespace: "default",
+				Labels: map[string]string{
+					"group": "group1",
+				},
+			},
+			Spec: monitoringv1.ServiceMonitorSpec{
+				Selector: metav1.LabelSelector{
+					MatchLabels: map[string]string{
+						"group": "group1",
+					},
+				},
+				Endpoints: []monitoringv1.Endpoint{
+					{
+						Port:     "web",
+						Interval: "30s",
+					},
+				},
+			},
+		}
+
+		if tc.scrapeClass != nil {
+			p.Spec.ScrapeClasses = []monitoringv1.ScrapeClass{*tc.scrapeClass}
+			if tc.scrapeClass.Default == nil {
+				monitor.Spec.ScrapeClass = ptr.To(tc.scrapeClass.Name)
+			}
+		}
+
+		cg := mustNewConfigGenerator(t, p)
+
+		cfg, err := cg.GenerateServerConfiguration(
+			context.Background(),
+			p.Spec.EvaluationInterval,
+			p.Spec.QueryLogFile,
+			p.Spec.RuleSelector,
+			p.Spec.Exemplars,
+			p.Spec.TSDB,
+			p.Spec.Alerting,
+			p.Spec.RemoteRead,
+			map[string]*monitoringv1.ServiceMonitor{
+				"monitor": &monitor,
+			},
+			nil,
+			nil,
+			nil,
+			&assets.Store{},
+			nil,
+			nil,
+			nil,
+			nil,
+		)
```

Comment:
```suggestion
		cfg, err := cg.GenerateServerConfiguration(
			context.Background(),
			p.Spec.EvaluationInterval,
			p.Spec.QueryLogFile,
			p.Spec.RuleSelector,
			p.Spec.Exemplars,
			p.Spec.TSDB,
			p.Spec.Alerting,
			p.Spec.RemoteRead,
			map[string]*monitoringv1.ServiceMonitor{"monitor": &monitor},
			map[string]*monitoringv1.PodMonitor{"monitor": &monitor},
			map[string]*monitoringv1.Probes{"monitor": &monitor},
			map[string]*monitoringv1alpha.ScrapeConfig{"monitor": &monitor},
			&assets.Store{},
			nil,
			nil,
			nil,
			nil,
		)
```

What do you think about doing a single test for all four, since the code for all 4 tests are 99% identical?

Reply:
I'd prefer have dedicated tests for each since IMHO this is more easy to maintain and evolve, but no strong opinion again.

I'm not sure if we will not have many checks on the code since there's no common spec against the monitors specs.

Reply:
@ArthurSens changed 😄